feat: (platform) Radio button group component Implementation

DeepakSap14 opened the pull request:
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/2121

https://github.com/SAP/fundamental-ngx/wiki/Platform:-RadioGroup-Technical-Design
#### Please provide a brief summary of this pull request.
Implementation of Radio Button Group component, which Groups many radio buttons with simple template signature.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

KevinOkamoto commented on libs/platform/src/lib/components/Forms/fdp-form.module.ts:
```diff
@@ -0,0 +1,30 @@
+        ReactiveFormsModule,
+        FdFormModule,
+        InlineHelpModule,
+        SelectModule,
```
Do we need to import all these modules (i.e. SelectModule, FdDatePickerModule, ComboboxModule) for the RadioButtonGroup component? I know we'll probably add those input elements later, but we should try to be minimalistic when submitting PRs.

KevinOkamoto commented on libs/platform/src/lib/components/Forms/radio-group/radio-group.component.html:
```diff
@@ -0,0 +1,31 @@
+        <fd-radio-button
+            fd-form-control
+            name="name"
+            [value]="None"
```
Is "None" a string value? If so, this will not work unless the string is placed in single quotes, like this:
```
[value]="'None'"
```

fkolar replied:
you dont use to use it this way. `[value]="None"` must even fail the compilation no?  you are using binding syntax. 
Also whey are u using directive `fd-form-control` ?

KevinOkamoto commented on libs/platform/src/lib/components/Forms/radio-group/radio-group.component.html:
```diff
@@ -0,0 +1,31 @@
+    <div fd-form-item *ngIf="hasNoValue">
+        <fd-radio-button
+            fd-form-control
+            name="name"
```
Are we hard-coding "name" to the string 'name'? I think you want property binding here:
```
[name]="name"
```

KevinOkamoto commented on libs/platform/src/lib/components/Forms/fdp-form.module.ts:
```diff
@@ -0,0 +1,30 @@
+    entryComponents: [RadioGroupComponent],
+    exports: [RadioGroupComponent]
+})
+export class FdpFormModule {}
```
I think we agreed to preface all our module names with `Platform`. Please change the name from `FdpFormModule` to `PlatformModule`

fkolar replied:
@KevinOkamoto Are you sure its Platform? I just told him its Fdp and I can pull out the thread.

KevinOkamoto replied:
Ah my bad - I guess we earlier decided on "Fdp". However all the other modules in the platform library are prefaced with "Platform", so instead of changing the names of the other modules which have already been published, let's name it "PlatformFormModule", so that we have consistent naming in our library, and from this point forward all platform modules will be prefaced with "Platform".

KevinOkamoto commented:
I checkout your branch, and tried running the documentation app, and saw these errors in the console:

<img width="1446" alt="Screen Shot 2020-03-05 at 7 59 59 AM" src="https://user-images.githubusercontent.com/48103491/75999753-8ca6f700-5eb7-11ea-8e1e-404451c95596.png">

fkolar commented on libs/platform/src/lib/components/Forms/form-control.ts:
```diff
@@ -0,0 +1,66 @@
+import {Observable} from 'rxjs';
+import {NgControl} from '@angular/forms';
+
+export abstract class FormFieldControl<T> {
```
We will do this in the next iteration after its merged

fkolar commented on libs/platform/src/lib/components/Forms/data-model.ts:
```diff
@@ -0,0 +1,65 @@
+ * custom label that is shown in the options.
+ */
+
+export type stateType = 'valid' | 'invalid' | 'warning' | 'default' | 'information';
```
also I saw similar constant in the core, we can simply use it, we dont need to redefine it here

fkolar commented on libs/platform/src/lib/components/Forms/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,77 @@
+import { BaseInput } from '../base.input';
+import { SelectItem } from '../data-model';
+
+export type stateType = 'valid' | 'invalid' | 'warning' | 'default' | 'information';
```
Similar as above for `statetype`. states are already defined in the core. can we use it?

DeepakSap14 commented on libs/platform/src/lib/components/form/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,282 @@
+                    button.inputElement.nativeElement.checked = true;
+                    this._selected = button;
+                }
+                // button.onChange.Pipe(takeUntil(this.destroy$)).subscribe(ev => this._selectedRadioButtonChanged(ev));
```
@fkolar , Radio button have not exposed change event. By using Onchange function i am not able to get button change event. please suggest here.

DeepakSap14 commented on libs/platform/src/lib/components/form/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,282 @@
+    set name(value: string) {
+        if (this.name !== value) {
+            this._name = value;
+            this._updateRadioButtonNames();
```
@fkolar ,as content is not initialised. I am not able to update names for content Radio buttons. please suggest

fkolar commented:
Since I was commenting on this mostly in the slack, have you addressed all the comments before I starts looking at this ? Also you might want to squash your commits. when you giving this out for a review..

DeepakSap14 commented:
> Since I was commenting on this mostly in the slack, have you addressed all the comments before I starts looking at this ? Also you might want to squash your commits. when you giving this out for a review..

i will take one more day to push the changes. i will let you know, if i push changes.
from next time i will squash the commits.

fkolar commented on libs/platform/src/lib/components/form/radio/radio.component.html:
```diff
@@ -0,0 +1,22 @@
+<ng-template #renderer>
+    <ng-container *ngTemplateOutlet="radio"></ng-container>
```
I have commented this before on slack, this is not addressed.
1. I was asking, why do you need this ng-template `radio` and extra ng-container?

fkolar commented on libs/platform/src/lib/components/form/radio/radio.component.spec.ts:
```diff
@@ -0,0 +1,24 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Remove this file as there are not tests and issue and it will be tested  as part of RB Group 

- Can you move Radio under RadioGroup? I know logically in case of CheckBox, it makes sense to have it as extra folder, but I dont see a usecase somebody will be using fdp-radio on its own (they could use fd-radio), but our added value is RB Group.

fkolar commented on libs/platform/src/lib/components/form/radio/radio.component.ts:
```diff
@@ -0,0 +1,175 @@
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    providers: [{ provide: FormFieldControl, useExisting: PlatformRadioButtonComponent, multi: true }]
+})
+export class PlatformRadioButtonComponent implements ControlValueAccessor {
```
Once again, you havent addressed thing that I said before. You asked me to comment things in slack and I have added comment for this. 

Check e.g. RB implementation for radio button in material design, and comment I said in slack

fkolar commented on libs/platform/src/lib/components/form/radio/radio.component.ts:
```diff
@@ -0,0 +1,175 @@
+export class PlatformRadioButtonComponent implements ControlValueAccessor {
+    /** name of radio button, will be set from radio button group name */
+    @Input()
+    get name(): any {
```
I am finishing my review as the same applies for name, its not addressed, Please come back when you address all, or at least you raise question.

I don't feel like copying over all what I said... Please next time, dont ask for input on slack..

DeepakSap14 replied:
i had not pushed changes for Radio button group. Now i have pushed.
yes slack is not a good place for comments.

fkolar commented on libs/platform/src/lib/components/form/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,219 @@
+     * values of radio buttons. type will be of SelectItem or String
+     */
+    @Input()
+    list: Array<SelectItem | string>;
```
What I mean by this list `list: Array<SelectItem | string>;` in relation to `CollectionBaseInput ` so to move this `@Input` into this parent, so we can use the same for CheckboxGroup. 
Since its CollectionBaseInput, it will have this `ist: Array<SelectItem | string>;`.

So you could remove it from here..

fkolar commented on libs/platform/src/lib/components/form/data-model.ts:
```diff
@@ -0,0 +1,33 @@
+/**
```
I though that this one is under: 

https://github.com/SAP/fundamental-ngx/blob/master/libs/platform/src/lib/domain/data-model.ts

You have moved this?

DeepakSap14 commented on libs/platform/src/lib/components/form/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,219 @@
+     * Type of Radio buttons.
+     */
+    @Input()
+    state: stateType = 'default';
```
if we do like this then it will diverse from fiori specification. and QA may raise more questions.

DeepakSap14 commented on libs/platform/src/lib/components/form/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,219 @@
+    ngAfterViewInit() {
+        // Mark this component as initialized in AfterViewInit
+        // because all views with contents will be initialised After this
+        setTimeout(() => {
```
I tried without Timeout using change detection detectchanges(), but on controller update for template driven form using onChange() (registerOnChange function), i am getting "expression has changed error".  I tried using "static: true", but this also did not help. i have spent 7-8 hours debugging this but could not resolve without Timeout().

fkolar commented on libs/platform/src/lib/components/form/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,223 @@
+    templateUrl: './radio-group.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class GroupRadioButtonComponent extends CollectionBaseInput implements AfterViewInit {
```
Rename to RadioGroupComponent

fkolar commented on libs/platform/src/lib/components/form/radio-group/radio-group.module.ts:
```diff
@@ -0,0 +1,13 @@
+    exports: [GroupRadioButtonComponent],
+    declarations: [GroupRadioButtonComponent]
+})
+export class FdpRadioButtonGroupModule {}
```
Rename to` PlatformRadioGroupModule`

fkolar commented on libs/platform/src/lib/components/form/radio-group/radio-group.component.ts:
```diff
@@ -0,0 +1,223 @@
+export class RadioGroupComponent extends CollectionBaseInput implements AfterViewInit {
+    /** value of selected radio button */
+    @Input()
+    get value(): any {
```
can you just do this instead:

```
@Input()
    get value(): any {
        return super.getValue();
    }

    set value(value: any) {
        super.setValue(value);
    }
```

You dont need to call writevalue right ?

fkolar commented on libs/platform/src/lib/components/form/radio-group/radio/radio.module.ts:
```diff
@@ -0,0 +1,12 @@
+    imports: [CommonModule, FdFormModule, RadioModule, FormsModule, ReactiveFormsModule],
+    exports: [RadioButtonComponent],
+    declarations: [RadioButtonComponent]
+})
```
We said that we dont need this module, radio will be part of RadioGroup. Unless it make sense to use this on its own. What do you think?

fkolar commented on libs/platform/src/lib/components/form/radio-group/radio/radio.component.ts:
```diff
@@ -0,0 +1,120 @@
+    @Input()
+    get state(): Status | stateType {
+        return this._state;
+    }
```
Who or what will be setting a state in case of RB?

DeepakSap14 replied:
changing the style for Error, which is _status in baseInput, It needs to be passed to Radio button. so i am doing [state]="status" in "radio-group.component.html".

DeepakSap14 replied:
have removed state input and getting it from Status of BaseInput.

fkolar commented:
@SAP/fundamental-denoland-team  check you check?

DeepakSap14 commented:
No change needed for 0.16.0 release. so merging it.

droshev commented:
@DeepakSap14 you should have checked with us before merging

DeepakSap14 commented:
@droshev "@SAP/fundamental-denoland-team" , was asked for review. i waited for 1 day for any comments. then i merged. 
Is any review pending, i will create separate PR, if it is needed.
